### PR TITLE
Change p4c bmv2 backend to support runtime array index

### PR DESCRIPTION
**Pull request by hesingh**
See https://github.com/p4lang/p4c/issues/2007

### CODE REVIEW

**Comment by hesingh** [comment]
> You need to add at least one test case, probably with an stf file to make sure that the code executes correctly.

Sure.  Right now, I am using https://github.com/IETF-Hackathon/p4-ipv6-switch-ml/blob/master/ipv6-switch-ml-bmv2.p4 as one test case.  This test also has a STF file.  I will also test simple_switch to see if the behavioral-model accepts the generated JSON for DEREFERENCE_STACK op.  I sent out the PR so that folks such as Antonin could take the diffs and see if the generated JSON looks ok.

**Comment by hesingh** [comment]
Here is JSON output due to code changes of this PR:

```
{
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["vector[0]", "e"]
            },
            {
              "type" : "field",
              "value" : [
                {
                  "type" : "header",
                  "value" : {
                    "value" : {
                      "op" : "DEREFERENCE_STACK",
                      "left" : {
                        "type" : "stack",
                        "value" : "pool_0"
                      },
                      "right" : {
                        "type" : "field",
                        "value" : ["hdr", "ml", "idx"]
                      }
                    },
                    "type" : "expression"
                  }
                }
              ]
            }
          ],
          "source_info" : {
            "filename" : "/home/hemant/p4-ipv6-switch-ml/ipv6-switch-ml-bmv2.p4",
            "line" : 109,
            "column" : 13,
            "source_fragment" : "hdr.vector[0].e = pool[hdr.ml.idx].val"
          }
}
```

**Comment by hesingh** [comment]
The P4 test code used for this PR does not work yet with latest behavioral-model and its simple_switch.  So far, I have been able to see that at least `behavioral-model/src/bm_sim/P4Objects.cpp` needs a change if the generated JSON is acceptable.  Let's first review the JSON output and then one can look into any changes to the behavioral-model code.

**Comment by hesingh** [comment]
The full JSON file output from code changes of this PR has been added to a new issue filed against the behavioral-model.  Please see 

https://github.com/p4lang/behavioral-model/issues/835

**Comment by mbudiu-vmw** [comment]
these tests are failing.

File: testdata/p4_16_samples/ipv6-switch-ml-bmv2.p4 [hunk 1]
```diff
@@ -49,14 +49,29 @@ parser MyParser(packet_in packet, out headers hdr, inout metadata_t meta,
     }
     state parse_udp {
         packet.extract(hdr.udp);
-        transition accept;	
+        transition parse_ml;
```
Comment (mbudiu-vmw) [inline testdata/p4_16_samples/ipv6-switch-ml-bmv2.p4]: in general we do not modify tests, but we add new ones.

**Comment by hesingh** [comment]
> these tests are failing.

This PR code is not complete yet because I got diverted to other priorities.  When the code is complete and tests pass, I will add a note to this PR.

**Comment by hesingh** [comment]
Another PR is handling the code changes of this PR. 
https://github.com/p4lang/p4c/pull/2139

**Pull request closed without merging**